File: #1717.maximum-score-from-removing-substrings.cpp
class Solution {
public:
    int maximumGain(string s, int x, int y) {
        
        string target = (x >= y) ? "ab" : "ba";
        stack<pair<char,int>> st;
        int result = 0;
        int n = s.size();
        vector<int> ind(n, 0);
        for (int i = 0; i < n; i++) {
            char ch = s[i];
            if (!st.empty()
                && st.top().first == target[0]
                && ch==target[1]) {
                result += std::max(x, y);
                ind[i] = 1;
                ind[st.top().second] = 1;
                st.pop();
            } else {
                st.push({ch, i});
            }
        }

        while (!st.empty()) {
            st.pop();
        }

        target = target == "ab" ? "ba" : "ab";

        for (int i = 0; i <n ; i++) {
            if (ind[i]) {
                continue;
            }

            char ch=s[i];
            if (!st.empty()
                && st.top().first == target[0]
                && ch == target[1]) {
                result += std::min(x, y);
                st.pop();
            } else {
                st.push({ch, i});
            }
        }
        return result;
    }
};

File: #1717.maximum-score-from-removing-substrings.ts
function maximumGain(s: string, x: number, y: number): number {
    let maxPoints = 0;

    const removePattern = (s: string, first: string, second: string, points: number): [string, number] => {
        let stack: string[] = [];
        let score = 0;

        for (let char of s) {
            if (stack.length && stack[stack.length - 1] === first && char === second) {
                stack.pop();
                score += points;
            } else {
                stack.push(char);
            }
        }
        return [stack.join(''), score];
    };

    if (x >= y) {
        let [intermediate, points] = removePattern(s, 'a', 'b', x);
        maxPoints += points;
        [, points] = removePattern(intermediate, 'b', 'a', y);
        maxPoints += points;
    } else {
        let [intermediate, points] = removePattern(s, 'b', 'a', y);
        maxPoints += points;
        [, points] = removePattern(intermediate, 'a', 'b', x);
        maxPoints += points;
    }
    return maxPoints;
};
